src/python/pants/backend/python/util_rules/pex_from_targets.py

```diff
@@ -244,13 +244,25 @@ async def pex_from_targets(
         # In requirement strings, Foo_-Bar.BAZ and foo-bar-baz refer to the same project. We let
         # packaging canonicalize for us.
         # See: https://www.python.org/dev/peps/pep-0503/#normalized-names
-        exact_req_projects = {
-            canonicalize_project_name(Requirement.parse(req).project_name) for req in exact_reqs
-        }
+
+        url_reqs = set()  # E.g., 'foobar@ git+https://github.com/foo/bar.git@branch'
```

Comment:
This PR adds a lot of randomness from all appearances. Lots of sets and set math. It seems the original order of the constraints and or requirements should be kept.

Reply:
I had the same thought, but the `PexRequirements` ctor sorts its argument. So we were already not preserving the original order, but we were (and still are) eliminating randomness.